pkg/rebalancer/storage/controller/service/rebalancer.go

```diff
@@ -1,47 +1,530 @@
-//
-// Copyright (C) 2019-2020 Vdaas.org Vald team ( kpango, rinx, kmrmt )
-//
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at
-//
-//    https://www.apache.org/licenses/LICENSE-2.0
-//
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-//
-
-// Package service manages the main logic of server.
 package service
 
 import (
 	"context"
+	"math"
 	"reflect"
+	"sync/atomic"
+	"time"
 
+	"github.com/vdaas/vald/internal/errgroup"
 	"github.com/vdaas/vald/internal/errors"
+	"github.com/vdaas/vald/internal/k8s"
+	"github.com/vdaas/vald/internal/k8s/configmap"
+	"github.com/vdaas/vald/internal/k8s/job"
+	mpod "github.com/vdaas/vald/internal/k8s/metrics/pod"
+	"github.com/vdaas/vald/internal/k8s/pod"
+	"github.com/vdaas/vald/internal/k8s/statefulset"
+	"github.com/vdaas/vald/internal/log"
+	"github.com/vdaas/vald/internal/safety"
+	"github.com/vdaas/vald/pkg/rebalancer/storage/controller/model"
+
+	// TODO: move to internal after internal/k8s refactoring
+	batchv1 "k8s.io/api/batch/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/conversion"
 )
 
+// Rebalancer represents the rebalancer interface.
 type Rebalancer interface {
-	Start(context.Context) (<-chan error, error)
+	Start(ctx context.Context) (<-chan error, error)
 }
 
 type rebalancer struct {
+	jobs           atomic.Value
+	jobName        string
+	jobNamespace   string
+	jobTemplate    atomic.Value
+	jobTemplateKey string // config map key of job template
+
+	agentName         string
+	agentNamespace    string
+	agentResourceType string // TODO: use custom type insteaf of string
+	pods              atomic.Value
+	podMetrics        atomic.Value
+
+	// job template config map
+	configmapName      string
+	configmapNamespace string
+
```

Comment:
maybe it is more clear?
```suggestion
	jobConfigmapName      string
	jobConfigmapNamespace string

```